text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,641 @@
+  batch, it would still be useful to capture related message-specific
+  attributes as part of a link which points to an invalid context. However,
+  according to the specification links pointing to an invalid context may be
+  ignored. For consistently reporting message-specific attributes on links, links
+  to invalid contexts should be allowed and supported.
+* [open-telemetry/opentelemetry-specification#3172](https://github.com/open-telemetry/opentelemetry-specification/issues/3172)
+  Currently the specification is unclear about whether relationships between
+  producer and consumer spans can be modelled via links, the wording suggests
+  that it should be a parent/child relationship. The wording in the specification
+  needs to make it clear that this can be a link too.
+
+## Examples
+
+This section contains a list of examples illustrating the use of the
+conventions outlined above. Green boxes denote spans that are required to
+exist in order to conform to those conventions. Other boxes denote spans that
+are not required and covered by the conventions, but are hopefully helpful in
+understanding how messaging spans can be integrated into an overall trace flow.
+Solid arrows denote parent/child relationships, dotted arrows denote link
+relationships.
+
+### Push-based scenarios
+
+A producer creates and publishes a single message, the single message is delivered to a consumer:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  PM1[Publish m1]
+  end
+  subgraph CONSUMER
+  DM1[Deliver m1]
+  end
+  PM1-. link .->DM1;
+
+  classDef normal fill:green
+  class PM1,DM1 normal
+  linkStyle 0 color:green,stroke:green
+```
+
+When consuming a single message, the "Deliver" spans can be parented to the creation context:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  PM1[Publish m1]
+  end
+  subgraph CONSUMER
+  DM1[Deliver m1]
+  end
+  PM1-. link .->DM1;
+  PM1-- parent -->DM1;
+
+  classDef normal fill:green
+  class PM1,DM1 normal
+  linkStyle 0,1 color:green,stroke:green
+```
+
+It is recommended to add spans for settlement operations. Those spans can
+either be created manually or via auto-instrumentation:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+    direction TB
+    PM1[Publish m1]
+  end
+  subgraph CONSUMER
+    direction TB
+    DM1[Deliver m1]-->S1[Settle m1]
+  end
+  PM1-. link .->DM1;
+  PM1-. link .->S1;
+
+  classDef normal fill:green
+  class PM1,DM1,S1 normal
+  linkStyle 0,1,2 color:green,stroke:green
+```
+
+A producer publishes a batch of messages, single messages are delivered to consumers:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  P[Publish]
+  end
+  subgraph CONSUMER
+  direction TB
+  DM1[Deliver m1]
+  DM2[Deliver m2]
+  end
+  P-. link .->DM1;
+  P-. link .->DM2;
+
+  classDef normal fill:green
+  class P,DM1,DM2 normal
+  linkStyle 0,1 color:green,stroke:green
+```
+
+When consuming a single message, the "Deliver" spans can be parented to the creation context:
```

Comment:
Again, here we aren't showing each message having a unique creation context.

Reply:
I added "Create" spans in the example.